### PR TITLE
feat: Adding Validation to Uno Extensions

### PR DESCRIPTION
**Pull request by nickrandolph**
GitHub Issue (If applicable): #981

## PR Type

What kind of change does this PR introduce?
- Feature

## What is the current behavior?

No support for validation in extensions

## What is the new behavior?

Adds support for validation

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [ ] Associated with an issue (GitHub or internal)

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.Extensions.Validation/IValidator.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿namespace Uno.Extensions.Validation;
+
+/// <summary>
+/// Defines an interface for a data validator.
+/// </summary>
+/// <typeparam name="T">Instance to validate</typeparam>
+public interface IValidator<in T>
```
Comment (dr1rrb) [inline src/Uno.Extensions.Validation/IValidator.cs:7]: Please see my comment on the discussion https://github.com/unoplatform/uno.extensions/discussions/324#discussioncomment-4380008 :)

File: src/Uno.Extensions.Validation/SystemValidator.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+		var succeeded = Validator.TryValidateObject(instance, context, results, true);
+
+		if (!succeeded && instance is INotifyDataErrorInfo _instance)
+		{
+			results = _instance?.GetErrors(null).OfType<ValidationResult>()?.ToList()
+				?? new List<ValidationResult>();
+		}
+
+		return new ValueTask<IEnumerable<ValidationResult>>(results.ToList());
```
Comment (dr1rrb) [inline src/Uno.Extensions.Validation/SystemValidator.cs:29]:
```suggestion
		return new ValueTask<IEnumerable<ValidationResult>>(results);
```

File: src/Uno.Extensions.Validation/Uno.Extensions.Validation.csproj [hunk 3]
```diff
@@ -0,0 +1,19 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+	<Import Project="..\tfms-non-ui.props" />
+
+	<PropertyGroup>
+		<Description>Validation extensions for Uno Platform, UWP and WinUI</Description>
+
+		<!--Temporary disable missing XML doc until fixed in the whole package-->
+		<WarningsNotAsErrors>$(WarningsNotAsErrors);CS1591</WarningsNotAsErrors>
```
Comment (dr1rrb) [inline src/Uno.Extensions.Validation/Uno.Extensions.Validation.csproj:8]: Remove this for new projects.

File: src/Uno.Extensions.Validation.Fluent/Uno.Extensions.Validation.Fluent.csproj [hunk 4]
```diff
@@ -0,0 +1,22 @@
+<Project Sdk="Microsoft.NET.Sdk">
+	<Import Project="..\tfms-non-ui.props" />
+
+	<PropertyGroup>
+		<Description>FluentValidation extensions for Uno Platform, UWP and WinUI</Description>
+
+		<!--Temporary disable missing XML doc until fixed in the whole package-->
+		<WarningsNotAsErrors>$(WarningsNotAsErrors);CS1591</WarningsNotAsErrors>
+
```
Comment (dr1rrb) [inline src/Uno.Extensions.Validation.Fluent/Uno.Extensions.Validation.Fluent.csproj:9]:
```suggestion
```

We must document new stuff :)

File: src/Uno.Extensions.Validation/Uno.Extensions.Validation.csproj (hunk 3 above)
Comment (dr1rrb) [reply src/Uno.Extensions.Validation/Uno.Extensions.Validation.csproj:8]:
```suggestion
```

File: src/Uno.Extensions.Validation.Fluent/Uno.Extensions.Validation.Fluent.csproj [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+		<!-- As we are InternalsVisibleTo, we disable some compatibility types that are not used by package itself. -->
+		<UnoExtensionsGeneration_DisableModuleInitializerAttribute>True</UnoExtensionsGeneration_DisableModuleInitializerAttribute>
```
Comment (dr1rrb) [inline src/Uno.Extensions.Validation.Fluent/Uno.Extensions.Validation.Fluent.csproj:11]:
```suggestion
		<!-- As we are InternalsVisibleTo, we disable some compatibility types that are not used by package itself. -->
		<UnoExtensionsGeneration_DisableModuleInitializerAttribute>True</UnoExtensionsGeneration_DisableModuleInitializerAttribute>
```

You should be able to remove this also. Insteed we should remove the `InternalsVisibleTo` if needed.

File: src/Uno.Extensions.Validation/Uno.Extensions.Validation.csproj [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+		<!-- As we are InternalsVisibleTo, we disable some compatibility types that are not used by package itself. -->
+		<UnoExtensionsGeneration_DisableModuleInitializerAttribute>True</UnoExtensionsGeneration_DisableModuleInitializerAttribute>
```
Comment (dr1rrb) [inline src/Uno.Extensions.Validation/Uno.Extensions.Validation.csproj:11]:
```suggestion

		<!-- As we are InternalsVisibleTo, we disable some compatibility types that are not used by package itself. -->
		<UnoExtensionsGeneration_DisableModuleInitializerAttribute>True</UnoExtensionsGeneration_DisableModuleInitializerAttribute>
```

Same here, you should be able to remove this.

**Review by dr1rrb — CHANGES_REQUESTED** [review]

**Comment by nickrandolph** [comment]
Replaced by https://github.com/unoplatform/uno.extensions/pull/1011

**Pull request closed without merging**